### PR TITLE
Feature/improve xml request error handling

### LINKED ISSUE #3425: XMLHttpRequest errors are not handled correctly
In the event of an error occurring during a XMLHttpRequest, `onreadystatechange` is called when
- `request.readyState` is `4` 
- `request.responseText` is `""`
- `request.status` is `0`

The following code that handles http requests does not handle this case, and can return a misleading error to the user when it occurs

https://github.com/ethereum/web3.js/blob/dee72e00883c5a23a47b6d3c1d27304ab1d4922e/packages/web3-providers-http/src/index.js#L87-L125

Let's pretend a machine has exhausted all of it's ephemeral ports right before web3.js tries to send a request. Here's what happens
1. `HttpProvider.prototype.send` is called
2. The request is crafted
3. `onreadystatechange` and `ontimeout` handlers are assigned
4. `request.send` is called with the payload
5. The request fails before it is even sent to the destination due to there being no ephemeral ports available 
6. `request.readyState` is set to `4`, `request.responseText` is set to `""`, `request.status` is set to `0` and the `onreadystatechange` handler is called
7. web3.js tries to parse `""` and throws an error that looks like this
```
  <rejected> Error: Invalid JSON RPC response: ""
      at Object.InvalidResponse (/home/liam/spells/web3.js/packages/web3-core-helpers/src/errors.js:42:16)
      at XMLHttpRequest.request.onreadystatechange (/home/liam/spells/web3.js/packages/web3-providers-http/src/index.js:108:32)
      at XMLHttpRequestEventTarget.dispatchEvent (/home/liam/spells/web3.js/packages/web3-providers-http/node_modules/xhr2-cookies/dist/xml-http-request-event-target.js:34:22)
      at XMLHttpRequest._setReadyState (/home/liam/spells/web3.js/packages/web3-providers-http/node_modules/xhr2-cookies/dist/xml-http-request.js:208:14)
      at XMLHttpRequest._onHttpRequestError (/home/liam/spells/web3.js/packages/web3-providers-http/node_modules/xhr2-cookies/dist/xml-http-request.js:349:14)
      at ClientRequest.<anonymous> (/home/liam/spells/web3.js/packages/web3-providers-http/node_modules/xhr2-cookies/dist/xml-http-request.js:252:61)
      at ClientRequest.emit (events.js:210:5)
      at ClientRequest.EventEmitter.emit (domain.js:475:20)
      at Socket.socketErrorListener (_http_client.js:406:9)
      at Socket.emit (events.js:210:5)
      at Socket.EventEmitter.emit (domain.js:475:20)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:80:21)

```
this error is misleading because it indicates that the node successfully received the request and actually returned an empty string in response, when in reality the request never even made it to the node and an empty string is just the default value set on `request` in an error scenario. 

I believe this may be the cause of some other issues people have opened in the past like https://github.com/ethereum/web3.js/issues/3370 and https://github.com/ethereum/web3.js/issues/1919

I think it makes sense to at least add an `onerror` handler to the `request`, and in the case that a request fails the error thrown indicates that instead of a misleading `InvalidResponse` error.

I'm working on a PR for this.

### PR DESCRIPTION
**Pull request by liamaharon**
## Description

Adds explicit handling for XMLHttpRequest errors.

When an XMLHttpRequest fails, instead of throwing `Error: Invalid JSON RPC response: ""` throw `Error: HTTP request failed`.

Unfortunately, accessing the specifics about what caused a request error when using `XMLHttpRequest` is not trivial. 

In future versions of web3.js maybe `XMLHttpRequest` could be substituted for [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API), which is built into modern browsers (so no increase to bundle size). As well making informative error handling really simple, a Fetch implementation would require far fewer lines of code and less complexity in the packages/web3-providers-http/src/index.js file.

Fixes #3425 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation. *Unsure where this would need to be documented?*
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran ```npm run dtslint``` with success and extended the tests and types if necessary.
- [x] I ran ```npm run test:unit``` with success.
- [ ] I have executed ``npm run test:cov`` and my test cases do cover all lines and branches of the added code. *Unsure if required to write tests for this change as there don't appear to be tests for similar logic branches, happy to if it's required just let me know*
- [x] I ran ```npm run build-all``` and tested the resulting file/'s from  ```dist``` folder in a browser.
- [x] I have updated the ``CHANGELOG.md`` file in the root folder. *Did I do this right?*
- [x] I have tested my code on the live network.

First time opening a PR here so let me know if I made any mistakes, happy to revise ✌️

### CODE REVIEW

File: packages/web3-providers-http/src/index.js [hunk 1]
```diff
@@ -111,6 +111,10 @@ HttpProvider.prototype.send = function (payload, callback) {
         }
     };
 
+    request.onerror = function() {
+        callback(errors.RequestFailed());
```
Comment (nivida) [inline packages/web3-providers-http/src/index.js:116]: It would probably be great to get our users some more information about the failed requests. :-)

Comment (liamaharon) [reply packages/web3-providers-http/src/index.js:116]: Agreed :) 

I wrote a little bit about this towards the end of the PR description, would you be open to having `XMLHttpRequest` being replaced with another library like `fetch` as part of this PR?

**Comment by wbt** [comment]
Would it be too much to ask to request an error number be included, instead of just an error string, to avoid upgrade issues like [this](https://github.com/ethereum/web3.js/issues/3425#issuecomment-619650271) with future string clarifications?

**Comment by GregTheGreek** [comment]
This would be nice to revive, @liamaharon do you mind updating your branch accordingly and we can take a look at it :)

**Comment by liamaharon** [comment]
@wbt I think this is sensible. The easiest (and laziest) way to implement this would be to add a prefix or suffix to all the error strings, for example something like `Error ({errCode}): {errMsg}` or `{errMsg} (code {errCode})`. This way you could match for the error code, rather than the error message in the string. Is this along the lines of what you're thinking?

@GregTheGreek sure, I've merged the latest `1.x` into my branch. Let me know if anything else needs doing, happy to work with you to move this forward.

**Comment by wbt** [comment]
@liamaharon 

> @wbt I think this is sensible. The easiest (and laziest) way to implement this would be to add a prefix or suffix to all the error strings, for example something like `Error ({errCode}): {errMsg}` or `{errMsg} (code {errCode})`. This way you could match for the error code, rather than the error message in the string. Is this along the lines of what you're thinking?

That still seems a bit hacky. In more professionalized packages, the errors thrown are objects rather than strings, with members "message" (the existing error message), "code" (a numeric code), and sometimes other fields.  I was thinking along those lines.

**Comment by uluhonolulu** [comment]
Please don't close it, the current error message is too confusing and requires very specific know-how to understand.

**Comment by wbt** [comment]
@uluhonolulu LOL; bots apparently don't read.

**Comment by uluhonolulu** [comment]
@GregTheGreek I would like to pick up this PR as @liamaharon asked me. The only problem is I've never worked on somebody else's PR's, so I don't know how to do it technically. Should I open a new PR or contribute to this one somehow? If the latter, can you assign it to me, please?

**Comment by GregTheGreek** [comment]
> @GregTheGreek I would like to pick up this PR as @liamaharon asked me. The only problem is I've never worked on somebody else's PR's, so I don't know how to do it technically. Should I open a new PR or contribute to this one somehow? If the latter, can you assign it to me, please?

I would branch off of this pr and open a new one!

**Comment by uluhonolulu** [comment]
OK, I'm on it

**Comment by jdevcs** [comment]
Closing this PR in favor of https://github.com/ChainSafe/web3.js/issues/3425#issuecomment-958010563

**Pull request closed without merging**